Rework installation instructions (#7434)
scikit-image/scikit-image: INSTALL.rst

Installation via pip and conda
------------------------------------------------------------------------------

.. _install-via-pip:

pip
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Prerequisites to a pip install: you must be able to use ``pip`` on
your command line to install packages.

We strongly recommend the use of a
`virtual environment
<https://towardsdatascience.com/virtual-environments-104c62d48c54?gi=2532aa12906#ee81>`_.

> They create a clean Python environment that does not interfere

Reviewer: ```suggestion
Creating a virtual environment is about creating a clean Python environment that does not interfere
```

Reviewer 2: Fixed.

> with the existing system installation, can be easily removed, and contain only

Reviewer: ```suggestion
with your existing system installation, that can be easily removed, and that contains only
```

Reviewer 2: I added the plural; I think the rest is OK as is.

  # Update pip
  python -m pip install -U pip

  # Install scikit-image
  python -m pip install -U scikit-image

> To access the example datasets from ``skimage.data``, use:

Reviewer 2: Will fix.

.. code-block:: sh

   python -m pip install -U scikit-image[data]

To install optional scientific Python packages that expand
``scikit-image``'s capabilities to include, e.g., parallel processing,
use:

.. code-block:: sh

    python -m pip install -U scikit-image[optional]

.. warning::

    Do not use the command ``sudo`` and ``pip`` together as ``pip`` may
    overwrite critical system libraries.

.. _install-via-conda:

conda
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

We recommend `miniforge <https://github.com/conda-forge/miniforge>`_, a minimal
distribution that makes use of `conda-forge <https://conda-forge.org>`_.
It installs Python and provides virtual environments.

Once you have your conda environment set up, install ``scikit-image`` with:

.. code-block:: sh

    conda install scikit-image

System package managers
------------------------------------------------------------------------------

> Using a package manager (``yum``, ``apt-get``, etc.) to install ``scikit-image``

Reviewer 2: Fixed.

or other Python packages is not your best option, since you're likely

> to get an older version. It also becomes harder to add other Python packages

Reviewer 2: Fixed.

not provided by the package manager.

Downloading all demo datasets
------------------------------------------------------------------------------

> Some of our example data (``skimage.data``) is hosted online and is

Reviewer 2: Fixed.

Installing scikit-image for contributors
========================================

> Your systems needs a:

Reviewer 2: Fixed.

- C compiler,
- C++ compiler, and
- a version of Python supported by ``scikit-image`` (see
  `pyproject.toml <https://github.com/scikit-image/scikit-image/blob/main/pyproject.toml#L14>`_).

First, `fork the scikit-image repository on GitHub <https://github.com/scikit-image/scikit-image/fork>`_.

> Then clone your fork locally and setup an ``upstream`` remote to point to the original scikit-image repository:

Reviewer: @stefanv here

- a version of Python supported by ``scikit-image`` (see
  `pyproject.toml <https://github.com/scikit-image/scikit-image/blob/main/pyproject.toml#L14>`_).

First, `fork the scikit-image repository on GitHub <https://github.com/scikit-image/scikit-image/fork>`_.
Then clone the fork locally and setup an ``upstream`` remote to point to the original scikit-image repository:

.. note::

    We use ``git@github.com`` below; if you don't have SSH keys setup, use
    ``https://github.com`` instead.

.. code-block:: sh

   git clone git@github.com:YOURUSERNAME/scikit-image
   cd scikit-image
   git remote add upstream git@github.com:scikit-image/scikit-image

> All commands below are run from within the new ``scikit-image`` directory.

Reviewer 2: new -> cloned?

  conda install -c conda-forge --file requirements/build.txt

  # Install scikit-image in editable mode. In editable mode,
  # scikit-image will be recompiled, as necessary, on import.
  spin install -v

.. tip::

    The above installs scikit-image into your environment, which makes
    it accessible to IDEs, IPython, etc.
    This is not strictly necessary; you can also build with:

    .. code-block:: sh

        spin build

    In that case, the library is not installed, but is accessible via
    ``spin`` commands, such as ``spin test``, ``spin ipython``, ``spin run``,
    etc.

Testing
-------

Run the complete test suite:

.. code-block:: sh

   spin test

Or run a subset of tests:

.. code-block:: sh

   # Run tests in a specific file
   spin test skimage/morphology/tests/test_gray.py

   # Run tests in the given directory
   spin test skimage/morphology

   # Run tests matching the given expression
   spin test -- -k local_maxima

Adding a feature branch
------------------------------------------------------------------------------

When contributing a new feature, do so via a feature branch.

First, fetch the latest source:

.. code-block:: sh

>    git checkout main

Reviewer: ```suggestion
   git switch main
```
@lagru please double-check

Reviewer 2: Fixed.

   git pull upstream main

Create your feature branch:

.. code-block:: sh

>    git checkout -b my-feature-name

Reviewer: ```suggestion
   git switch --create my-feature-name
```
Even if I'm old-style and I personally still use `git checkout -b` :grin:

  # Install development dependencies
  conda install -c conda-forge --file requirements/default.txt
  conda install -c conda-forge --file requirements/test.txt

>   conda install -c conda-forge --file requirements/developer.txt

Reviewer 2: Check if works

  pip install -r requirements.txt
  # Install build dependencies of scikit-image
  pip install -r requirements/build.txt

>   # Install scikit-image in editable mode, which will make sure that

Reviewer 2: You should make clear that this needs to happen only once.

  # scikit-image is recompiled if necessary on import

>   spin install -v

Reviewer 2: I'd split the flow here between everything up to install, and then, separately, show a few examples of what you can do with it.

Also, match the conda version to what is recommended here.

Remove mentions of Anaconda, we should only recommend miniforge.

  # Test your installation
  spin test

>   # Build docs

Reviewer: Also, we can't build docs (a priori) if we haven't installed the docs dependencies first:
```suggestion
  # Build docs
  pip install -r requirements/docs.txt
```

Author: I think you actually can. `spin docs` should actually take care of that...

Reviewer: Well, I couldn't when I tried the other day. I had to `pip install -r requirements/docs.txt` first. That's the only reason why I added this (off-topic) suggestion. :wink:

Author: Oh, you are right, good catch! `spin docs` actually isn't installing deps since https://github.com/scikit-image/scikit-image/pull/6940. That PR doesn't explain why we added the `--install-deps` option though and turned it off by default. And I also don't remember. :thinking:

Author: Adressed in 8f9aeaf59618cf3d34d1856656145ed1e4555985.

Reviewer: Thanks, Lars!

You can use ``scikit-image`` with the basic requirements listed above, but some
functionality is only available with the following installed:

* `Matplotlib <https://matplotlib.org>`__

>     Used in various functions, e.g., for drawing, segmenting, reading images.

Reviewer: ```suggestion
  Used in various functions, e.g., for drawing, segmenting, reading images.
```
Based on the following items?

* `Dask <https://dask.org/>`__
  The ``dask`` module is used to parallelize certain functions.

More rarely, you may also need:

* `PyAMG <https://pyamg.org/>`__
  The ``pyamg`` module is used for the fast ``cg_mg`` mode of random
  walker segmentation.

* `Astropy <https://www.astropy.org>`__
  Provides FITS I/O capability.

* `SimpleITK <http://www.simpleitk.org/>`__
  Optional I/O plugin providing a wide variety of `formats <https://itk.org/Wiki/ITK_File_Formats>`__.

>   including specialized formats using in medical imaging.

Reviewer: ```suggestion
  including specialized formats used in biomedical imaging.
```
